module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +217,111 @@ func TestInvalidateEntity(t *testing.T) {
+// LRUEjection, NoEjection and RandomEjection are tested. RandomEjection doesn't allow to provide a final state of the pool to check.
+func TestAddAndRemoveEntities(t *testing.T) {
+	for _, tc := range []struct {
+		limit               uint32       // capacity of the pool
+		entityCount         uint32       // total entities to be stored
+		ejectionMode        EjectionMode // ejection mode
+		numberOfOperations  int
+		probabilityOfAdding float32
+	}{
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        LRUEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        NoEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        RandomEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.numberOfOperations, tc.probabilityOfAdding)
+		})
+	}
+}
+
+// Indexes contains an index in the pool of an inserted entity and an index of this entity
+// in the array of entities created by the EntityListFixture.
+type Indexes struct {
+	indexInPool        EIndex
+	indexInEntitiesArr int
+}
+
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, numberOfOperations int, probabilityOfAdding float32) {
+	// create and log the seed.
+	testSeed := time.Now().UnixNano()
+	rand.Seed(testSeed)
+	fmt.Println("Seed used for this testAddRemoveEntities execution", testSeed)
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// an array of random owner Ids.
+	ownerIds := make([]uint64, entityCount, entityCount)
+	// generate ownerId to index in the entities arrray.
+	for i := 0; i < int(entityCount); i++ {
+		ownerIds[i] = rand.Uint64()
+	}
+	// this map maintains entities currently stored in the pool.
+	var addedEntities map[flow.Identifier]Indexes
+	addedEntities = make(map[flow.Identifier]Indexes)
+
+	for i := 0; i < numberOfOperations; i++ {
+		// choose between Add and Remove with a probability of probabilityOfAdding and 1-probabilityOfAdding respectively.
+		if rand.Float32() < probabilityOfAdding || len(addedEntities) == 0 {
+			// adding an entity
+			entityToAdd := rand.Intn(int(entityCount))
+			// check that entity is not already inserted.
+			_, found := addedEntities[entities[entityToAdd].ID()]
+			if !found {
+				indexInThePool, _, ejectedEntity := pool.Add(entities[entityToAdd].ID(), entities[entityToAdd], uint64(ownerIds[entityToAdd]))
+				if indexInThePool != InvalidIndex {
+					addedEntities[entities[entityToAdd].ID()] = Indexes{indexInThePool, entityToAdd}
+				}
```

Comment:
Unless the ejection mode is set to `NoEjection` _and_ `len(addedEntities) > limit`, the `indexInThePool` should never be `InvalidIndex`. Please add that check.

Reply:
Thank you, will do

Reply:
Refactored in my proposed changes: https://github.com/onflow/flow-go/pull/4470#discussion_r1263014762

Reply:
Please consider some minor changes. Please let me know if you agree with including them in your proposal before I submit it.
1.				
```
if ejectionMode != NoEjection || len(addedEntities) < int(limit) {
   // when there is an ejection mode in place, and the pool is not full, the index should be valid.
    require.NotEqual(t, InvalidIndex, indexInThePool)
}
```
the comment here is not correct as you can get to this line with ```ejectionMode== NoEjection``` and ```len(addedEntities) < int(limit)``` thus should be 
// when there is an ejection mode in place, OR the pool is not full, the index should be valid.

2.
```
if ejectionMode != NoEjection && len(addedEntities) >= int(limit) {
    // when there is an ejection mode in place, the ejected entity should be valid.
    require.NotNil(t, ejectedEntity)
}
```
If ```len(addedEntities) > limit``` then the test has to be stoped with a failure as a number of currently added entities cannot exceed the pool capacity. I would suggest to stick to ```==``` condition.

Reply:
Thanks for the suggestions: 
1. Agree: `// when there is an ejection mode in place, OR the pool is not full, the index should be valid.`
2. You are right, then let's consider it as a sanity check like following: 
     ```go
   require.LessOrEqual(t, len(addedEntities), int(limit), "pool should not contain more elements than its limit")
	if ejectionMode != NoEjection && len(addedEntities) == int(limit) {
		// when there is an ejection mode in place, the ejected entity should be valid.
		require.NotNil(t, ejectedEntity)
	}
     ```
 3. Agree, `toAddIndex` serves a better comprehension.

Reply:
Thx, let me implement all 3 right now.

Reply:
implemented, closing the thread.